cpu/sam0_common/periph/sercom.c

```diff
@@ -0,0 +1,404 @@
+    }
+#if defined(CPU_COMMON_SAML21)
+    else {
+        GCLK->PCHCTRL[SERCOM5_GCLK_ID_CORE].reg = (GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(gclk));
+    }
+#endif /* CPU_COMMON_SAML21 */
+#endif
+}
+
+/**
+ * @brief   Disable peripheral clock for given SERCOM device
+ *
+ * @param[in] sercom    SERCOM device
+ */
+static void _sercom_clk_disable(sercom_t sercom)
+{
+#if defined(CPU_COMMON_SAMD21)
+    PM->APBCMASK.reg &= ~(PM_APBCMASK_SERCOM0 << sercom);
+#elif defined (CPU_COMMON_SAMD5X)
+    if (sercom < 2) {
+        MCLK->APBAMASK.reg &= ~(1 << (sercom + 12));
+    } else if (sercom < 4) {
+        MCLK->APBBMASK.reg &= ~(1 << (sercom + 7));
+    } else {
+        MCLK->APBDMASK.reg &= ~(1 << (sercom - 4));
+    }
+#else
+    if (sercom < 5) {
+        MCLK->APBCMASK.reg &= ~(MCLK_APBCMASK_SERCOM0 << sercom);
+    }
+#if defined (CPU_COMMON_SAML21)
+    else {
+        MCLK->APBDMASK.reg &= ~(MCLK_APBDMASK_SERCOM5);
+    }
+#endif /* CPU_COMMON_SAML21 */
+#endif
+}
+
+/**
+ * @brief   Wait for a previous reset to be completed
+ */
+static void _wait_for_reset_to_complete(Sercom *dev)
+{
+#ifdef SERCOM_SPI_STATUS_SYNCBUSY
+    while (dev->SPI.STATUS.reg & SERCOM_SPI_STATUS_SYNCBUSY) {}
+#else
+    while (dev->SPI.SYNCBUSY.reg) {}
+#endif
+
+    while (dev->SPI.CTRLA.reg & SERCOM_SPI_CTRLA_SWRST) {}
+}
+
+void sercom_acquire(sercom_t sercom, const sercom_conf_t *conf,
+                    sercom_irq_cb_t irq_cb, void *irq_arg)
+{
+    assume((unsigned)sercom < SERCOM_NUMOF);
+    mutex_lock(&_locks[sercom]);
+    Sercom *dev = sercom_get_baseaddr(sercom);
+    DEBUG("[sercom] acquiring SERCOM%d @ %p\n", (unsigned)sercom, (void *)dev);
+
+    /* track state to detect possible double frees */
+    if (IS_ACTIVE(DEVELHELP)) {
+        bf_set_atomic(_sercoms_acquired, sercom);
+    }
+
+    _cbs[sercom] = irq_cb;
+    _args[sercom] = irq_arg;
+
+    _sercom_clk_enable(sercom, conf->gclk);
+
+    /* A potential previous call to sercom_release() did not wait for the reset
+     * to be complete to not waste time. But we need to do so now, in case it
+     * is not yet finished anyway */
+    _wait_for_reset_to_complete(dev);
+
+    /* SERCOM should now be reseted: Either, because this is the first
+     * time we acquire it or because sercom_release() did a reset of the
+     * SERCOM */
+    assume(!dev->SPI.CTRLA.reg);
+
+    /* Note: The BAUD register is either 8 bit (in case of SPI/I2C) or 16 bit
+     *       wide (in case of UART). Due to conf->baud being in little endian,
+     *       valid values for conf->baud for SPI/I2C mode will write zeros to
+     *       the part exceeding the 8 bit BAUD field. So while this write may
+     *       write to addresses that have no meaning, it happens to just work
+     *       in all modes.
+     */
+    dev->USART.BAUD.reg = conf->baud;
+    dev->USART.CTRLA.reg = conf->ctrla;
+    dev->USART.CTRLB.reg = conf->ctrlb;
+
```

Comment:
I would like to see CTRLC here

Reply:
Why? It is currently unused and would add overhead. We can still add it when we start to use it.

Reply:
I use it: with a untypical configuration on top of SERCOM, until now I avoided any RIOT to configure it ( and use the registers directly), because RIOT does not abstract SERCOM, with this PR it would be for the incompleteness.

I know this is 'sercom_internal.h' -- but Why did you name the header that way? Of cause not every MCU has a SERCOM but it is quiet standard on the SAM.

Reply:
I think my design of passing over the control regs was actually a stupid idea. I have changed the signature of `sercom_acquire()` to just take the gclk source and clock the SERRCOM, but let the control and baud register programming be performed by the caller, who knows best what registers need programming and e.g. how large the BAUD reg is.

Sadly, this somehow caused a regression in I2C. Maybe due to timing only, but I'll need to fix this first.

Reply:
I found out the culprit: In `master` the test is not testing I2C. And I2C hangs due to missing external pull ups.

It turns out that my logic analyzer does have pull ups enabled by default, so monitoring the I2C signals "fixes" the issue. I did not tested this before without the logic analyzer attached to I2C and I2C enabled. So the regression was caused by removing the external pull ups (in form of the logic analyzer), not but the code change.

At least for [SAM L21](https://onlinedocs.microchip.com/oxy/GUID-24121E3C-B8D3-4957-90D7-3EE3914A3F39-en-US-4/GUID-923E3852-43A4-4E32-BB38-09FEF4C50A2D.html) this is expected:

> Note: When the I2C is enabled, internal pull-up resistors are not available. External pull-up resistors are required for proper function.